python/sources/simple-csv-new/connector/source_connector.py

```diff
@@ -0,0 +1,178 @@
+from quixstreams.models.types import MessageKey, MessageValue
+
+
+logger = logging.getLogger(__name__)
+
+__all__ = ("CsvSourceConnector",)
+
+
+class CsvConsumer(SourceConsumer):
+    """
+    A thin wrapper around csv.DictReader for lazy row-based reading of a CSV file
+    with a header.
+    """
+
+    def __init__(
+        self,
+        path: str,
+        key_parser: Optional[Union[str, Callable[[dict], MessageKey]]] = None,
+        value_parser: Optional[Callable[[dict], MessageValue]] = json.dumps,
+        **csv_reader_kwargs,
+    ):
+        """
+        :param path: file path to a .csv file
+        :param key_parser: how a key is generated for a given data row. Options:
+            - None; no key will be generated (default)
+            - A string, representing a column name; uses that row's column value.
+            - A function that handles the row as a dict.
+        :param value_parser: A function for generating resulting message value for the
+            row. Default: json.dumps
+        :param csv_reader_kwargs: kwargs passthrough to csv.DictReader
+        """
+        self._path = path
+        self._key_parser = key_parser
+        self._value_parser = value_parser
+        self._file = None
+        self._reader = None
+        self._reader_kwargs = csv_reader_kwargs
+
+    def _as_message(self, data: dict) -> KafkaMessage:
+        """Convenience method to create a kafka-formatted message"""
+        return KafkaMessage(
+            key=self.key_parser(data),
+            value=self.value_parser(data),
+            headers=None
+        )
+
+    def _consume(self) -> Iterator[dict]:
+        """Lazy row reading of the CSV to avoid loading it all in memory"""
+        for row in self._reader:
+            yield row
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.stop()
+
+    def start(self):
+        if not self._file:
+            self._file = open(self._path, "r")
+            self._reader = csv.DictReader(self._file, **self._reader_kwargs)
+            logger.info(f"CSV Consumer beginning reading file: {self._path}")
+
+    def stop(self):
+        if self._file:
+            logger.info(f"CSV Consumer has stopped reading file: {self._path}")
+            self._file.close()
+            self._file = None
+        self._reader = None
+
+    def key_parser(self, data: dict) -> Optional[MessageKey]:
+        """
+        :param data: CSV row data as a dict
+
+        :return: a MessageKey type or None
+        """
+        if self._key_parser:
+            if isinstance(self._key_parser, str):
+                return data[self._key_parser]
+            return self._key_parser(data)
+
+    def value_parser(self, data: dict) -> Optional[MessageValue]:
+        """
+        :param data: CSV row data as a dict
+
+        :return: a MessageValue type or None
+        """
+        return self._value_parser(data)
```

Comment:
Again, lets have this as default implementation in base class.

Reply:
Not sure what you mean in this case, unless you mean we should assume the end user is always going to pass some sort of value parsing function?

I didn't want to restrict how exactly a "parser" should work, only that some sort of value parsing must happen in some capacity (because...of course!). 

In the case of a CSV it's fairly simple (especially since each row reads as a dictionary), but I didn't want to make assumptions about how value extraction should work for other sources...kept it as generic as possible and leave the implementation up to the specific source. Like, maybe a function from the user isn't necessary at all for some sources and the user should just pass two arguments or something, IDK.